source/d3d11/d3d11_device_context.hpp

```diff
@@ -189,8 +190,24 @@ struct D3D11DeviceContext : ID3D11DeviceContext3
 	virtual void STDMETHODCALLTYPE GetHardwareProtectionState(BOOL *pHwProtectionEnable) override;
 	#pragma endregion
 
+	// local methods
+	void set_active_depthstencil(ID3D11DepthStencilView* pDepthStencilView);
+	void log_drawcall(UINT vertices);
+	void clear_drawcall_stats();
+	void log_drawcall_stats_in_tracker();
+	depth_counter_tracker& get_depth_counter_tracker();
+	
+	// fields
 	LONG _ref = 1;
+	UINT _drawcalls = 0;
+	UINT _vertices = 0;
+	UINT _active_depthstencil_drawcalls = 0;
+	UINT _active_depthstencil_vertices = 0;
+	UINT _total_drawcalls = 0;
+	UINT _total_vertices = 0;
```

Comment:
I don't think `_active_depthstencil_drawcalls`, `_active_depthstencil_vertices`, `_total_drawcalls` and `_total_vertices` are needed, since the same information is duplicated in `_depthstencil_counters`, so it can be just stored there directly?

Reply:
No, that's not how it works :) 
_active_* counters are for the currently set depthstencil on the context. See the following flow (pseudocode):
- context.Draw() -> no active depthstencil, nothing is counted
- context.setdepthstencil(X) -> the active depthstencil is set to X in context.
- context.Draw() -> has active depthstencil, so incs _active_depthstencil_drawcalls and adds vertices to _active_depthstencil_vertices. 
- context.Draw() -> see previous
- context.Draw() -> see previous
- context.setdepthstencil(Y) -> has a current active depthstencil of X. It will add X with the current values for the _active_* counters in the context's drawcall tracker. Then sets Y as the active depthstencil
etc. 
At the end it also makes sure it collects the last drawcalls for the last active depthstencil, when the tracker is obtained. 

`_total_drawcalls` and `_total_vertices` are the absolute totals, so also the drawcalls which were executed when no depthstencil was active (see `D3D11DeviceContext::log_drawcall`). So all counters are needed and have potentially different values. 

When the context C1 is a deferred context, the drawcall tracker is stored with the commandlist created from it. When the commandlist is then executed on another context C2 the drawcall tracker stats of C1 (which are associated with the commandlist) are merged with the one in C2. This then all bubbles down to all stats being merged eventually in the tracker in the immediatecontext. That is, the ones which were actually executed (e.g. commandlist which was tossed away isn't merged. Commandlist which was executed 10 times is merged 10 times)